gunrock/app/lp/lp_problem.cuh

```diff
@@ -0,0 +1,511 @@
+
+/**
+ * @file
+ * lp_problem.cuh
+ *
+ * @brief GPU Storage management Structure for BFS Problem Data
+ */
+
+#pragma once
+
+#include <gunrock/app/problem_base.cuh>
+
+namespace gunrock {
+namespace app {
+namespace lp {
+
+/**
+ * @brief  Speciflying parameters for BFS Problem
+ * @param  parameters  The util::Parameter<...> structure holding all parameter
+ * info \return cudaError_t error message(s), if any
+ */
+cudaError_t UseParameters_problem(util::Parameters &parameters) {
+  cudaError_t retval = cudaSuccess;
+
+  GUARD_CU(gunrock::app::UseParameters_problem(parameters));
+
+  GUARD_CU(parameters.Use<bool>(
+      "mark-pred",
+      util::OPTIONAL_ARGUMENT | util::MULTI_VALUE | util::OPTIONAL_PARAMETER,
+      false, "Whether to mark predecessor info.", __FILE__, __LINE__));
+
+  return retval;
+}
+
+/**
+ * @brief Breadth-First Search Problem structure
+ * @tparam _GraphT  Type of the graph
+ * @tparam _LabelT  Type of labels used in sssp
+ * @tparam _ValueT  Type of per-vertex distance values
+ * @tparam _FLAG    Problem flags
+ */
+template <typename _GraphT, typename _LabelT = typename _GraphT::VertexT,
+          typename _ValueT = typename _GraphT::ValueT,
+          ProblemFlag _FLAG = Problem_None>
+struct Problem : ProblemBase<_GraphT, _FLAG> {
+  typedef _GraphT GraphT;
+  static const ProblemFlag FLAG = _FLAG;
+  typedef typename GraphT::VertexT VertexT;
+  typedef typename GraphT::SizeT SizeT;
+  typedef typename GraphT::CsrT CsrT;
+  typedef typename GraphT::CscT CscT;
+  typedef typename GraphT::GpT GpT;
+  typedef _LabelT LabelT;
+  typedef _ValueT ValueT;
+  typedef unsigned char MaskT;
+
+  typedef ProblemBase<GraphT, FLAG> BaseProblem;
+  typedef DataSliceBase<GraphT, FLAG> BaseDataSlice;
+
+  // Helper structures
+  /**
+   * @brief Data slice structure containing BFS-specific data on indiviual GPU
+   */
+  struct DataSlice : BaseDataSlice {
+    // TODO remove unneccesary fields from here
+    // util::Array1D<SizeT, VertexT> original_vertex;
+    util::Array1D<SizeT, LabelT>
+        labels;  // labels to mark latest iteration the vertex been visited
+    util::Array1D<SizeT, LabelT>
+    old_labels;
+    util::Array1D<SizeT, VertexT> preds;       // predecessors of vertices
+    util::Array1D<SizeT, VertexT> temp_preds;  // predecessors of vertices
+    util::Array1D<SizeT, SizeT> vertex_markers[2];
+    util::Array1D<SizeT, VertexT> unvisited_vertices[2];
+    util::Array1D<SizeT, SizeT, util::PINNED> split_lengths;
+    util::Array1D<SizeT, VertexT> local_vertices;
+    util::Array1D<SizeT, MaskT> visited_masks;
+    util::Array1D<SizeT, MaskT> old_mask;
+    util::Array1D<SizeT, MaskT *> in_masks;
+    // need two arrays
+    // store data array containing labels
+    // store segment information
+    // store the count of segments
+    // max segment size is frontier size
+    // max data array is sum of top frontier size neighbours lengths
```

Comment:
If these comments are describing the following arrays, could you place them next to the array which they describe?